File: generate_pwa_icons.py
#!/usr/bin/env python3
"""
Генерирует PNG-иконки для PWA из одного исходного PNG (Pillow).
Использование: python generate_pwa_icons.py <путь_к_исходному_png>
"""
import sys
from pathlib import Path

from PIL import Image

# Куда класть иконки
PUBLIC_DIR = Path(__file__).resolve().parent / "frontend" / "public"

SIZES = [
    ("apple-touch-icon-180x180.png", 180),
    ("pwa-192x192.png", 192),
    ("pwa-512x512.png", 512),
    ("pwa-maskable-512x512.png", 512),
]


def main():
    if len(sys.argv) < 2:
        print("Использование: python generate_pwa_icons.py <путь_к_png>")
        sys.exit(1)
    src_path = Path(sys.argv[1])
    if not src_path.exists():
        print(f"Файл не найден: {src_path}")
        sys.exit(1)
    PUBLIC_DIR.mkdir(parents=True, exist_ok=True)
    img = Image.open(src_path).convert("RGBA")
    for name, size in SIZES:
        out = img.resize((size, size), Image.Resampling.LANCZOS)
        out.save(PUBLIC_DIR / name, "PNG")
        print(f"Записано: {name}")
    print("Готово.")


if __name__ == "__main__":
    main()

File: frontend/scripts/generate-pwa-icons.mjs
#!/usr/bin/env node
/**
 * Генерирует PNG иконки для PWA из public/favicon.svg
 * Требует: npm i -D @resvg/resvg-js
 */
import { readFile, writeFile, mkdir } from 'fs/promises'
import { join, dirname } from 'path'
import { fileURLToPath } from 'url'
import { Resvg } from '@resvg/resvg-js'

const __dirname = dirname(fileURLToPath(import.meta.url))
const publicDir = join(__dirname, '..', 'public')
const svgPath = join(publicDir, 'favicon.svg')

const sizes = [
  { name: 'apple-touch-icon-180x180.png', size: 180 },
  { name: 'pwa-192x192.png', size: 192 },
  { name: 'pwa-512x512.png', size: 512 },
  { name: 'pwa-maskable-512x512.png', size: 512, maskable: true },
]

async function main() {
  const svg = await readFile(svgPath)
  for (const { name, size, maskable } of sizes) {
    const renderSize = maskable ? Math.round(size * 0.8) : size
    const resvg = new Resvg(svg, {
      fitTo: { mode: 'width', value: renderSize },
      background: '#ffffff',
    })
    const pngData = resvg.render()
    const pngBuffer = pngData.asPng()

    if (maskable) {
      const { Resvg: R2 } = await import('@resvg/resvg-js')
      const offset = Math.round((size - renderSize) / 2)
      const wrapSvg = `<svg xmlns="http://www.w3.org/2000/svg" width="${size}" height="${size}">
        <rect width="${size}" height="${size}" fill="white"/>
        <image x="${offset}" y="${offset}" width="${renderSize}" height="${renderSize}"
               href="data:image/png;base64,${pngBuffer.toString('base64')}"/>
      </svg>`
      const r2 = new R2(Buffer.from(wrapSvg), { fitTo: { mode: 'width', value: size } })
      await writeFile(join(publicDir, name), r2.render().asPng())
    } else {
      await writeFile(join(publicDir, name), pngBuffer)
    }
    console.log('Written', name)
  }
  console.log('Done.')
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})
